### PR TITLE
Added option to configure if/else brace style

### LINKED ISSUE #884: Option to put `else` on newline
As of rustfmt v0.3.0 (installed via cargo install --branch master), single_line_if_else does not do anything, or does not do what I would expect it to. Regardless of whether it it set to true or false, else and else if expressions are always formatted like so:

if condition {
//statements
} else {
//more statements
}

If single_if_else is not the configuration option to control this behaviour, then there should be a flag added to control this behaviour (whether else and else if are placed on the same line as the closing brace of the previous expression in the chain). Many coding styles prefer else and else if on the next line following a closing brace.

### PR DESCRIPTION
**Pull request by flutterhash**
Resolves #884, #946

### CODE REVIEW

**Comment by marcusklaas** [comment]
Thanks for the PR!

Could you add some tests for the new styles? [Contributing.md](https://github.com/rust-lang-nursery/rustfmt/blob/master/Contributing.md) has a short guide on adding tests with custom options.

**Comment by flutterhash** [comment]
Sure thing, sorry about the missing tests.

**Comment by flutterhash** [comment]
Tests have been added and the changes are confirmed to work. Please let me know if there are any other outstanding problems.

File: src/lib.rs [hunk 1]
```diff
@@ -155,6 +155,23 @@ impl Indent {
         }
         indent
     }
+
+    pub fn to_string_no_align(&self, config: &Config) -> String {
```
Comment (marcusklaas) [inline src/lib.rs]: This functionality already exists. You can use `Indent::to_string` for this :-)

File: src/expr.rs [hunk 2]
```diff
@@ -763,8 +769,12 @@ fn rewrite_if_else(context: &RewriteContext,
                                 offset,
                                 false)
             }
-            _ => else_block.rewrite(context, width, offset),
+            _ => {
+                last_in_chain = true;
+                else_block.rewrite(context, width, offset)
+            }
         };
+        let last_in_chain = last_in_chain;
```
Comment (marcusklaas) [inline src/expr.rs]: What's going on here?

File: src/expr.rs [hunk 3]
```diff
@@ -781,10 +791,26 @@ fn rewrite_if_else(context: &RewriteContext,
...
         let after_else_comment = extract_comment(after_else, &context, offset, width);
 
+        let between_sep = match context.config.else_if_brace_style {
+            ElseIfBraceStyle::AlwaysNextLine |
+            ElseIfBraceStyle::ClosingNextLine => alt_separator.as_str(),
+            _ => " ",
+        };
+        let after_sep = match context.config.else_if_brace_style {
+            ElseIfBraceStyle::AlwaysNextLine => {
+                if last_in_chain {
```
Comment (marcusklaas) [inline src/expr.rs]: Nit: you could use this as a guard for the `ElseIfBraceStyle::AlwaysNextLine` pattern.

**Comment by marcusklaas** [comment]
This looks pretty good! Nice tests, too :-)

r+ with the comments addressed.

File: src/lib.rs (hunk 1 above)
Comment (flutterhash) [inline src/lib.rs]: The behaviour is subtly different. It does not add on the extra alignment to the "if" statement, only the block indent. Unless you believe it's preferable to clone() the config passed into rewrite_if_else and to modify it to have an alignment of 0.

I did this to resolve an issue with Allman and Stroupstrup styles having bizarrely aligned braces for "let x = if ..." statements, where the opening brace would line up with the if, but the closing brace would line up with the "let".

File: src/expr.rs (hunk 2 above)
Comment (flutterhash) [inline src/expr.rs]: This is to remove the mutabiity of last_in_chain when it's no longer unnecessary. I am uncertain if this is desirable idiomatic Rust.

File: src/expr.rs (hunk 3 above)
Comment (flutterhash) [inline src/expr.rs]: I do not understand what you mean by "guard" in this context.

File: src/lib.rs (hunk 1 above)
Comment (flutterhash) [inline src/lib.rs]: Nevermind, I'm an idiot, I can just copy the indent.

**Comment by flutterhash** [comment]
I have also added loop brace options with tests.

File: src/expr.rs (hunk 3 above)
Comment (marcusklaas) [inline src/expr.rs]: You could write this as:
```rust
match context.config.else_if_brace_style {
    ElseIfBraceStyle::AlwaysNextLine if last_in_chain => alt_block_sep,
    _ => " ",
}
```
The extra condition on the first pattern is called the guard.

File: src/expr.rs [hunk 4]
```diff
@@ -781,10 +797,26 @@ fn rewrite_if_else(context: &RewriteContext,
                                else_block.span.lo);
         let after_else_comment = extract_comment(after_else, &context, offset, width);
 
+        let between_sep = match context.config.else_if_brace_style {
+            ElseIfBraceStyle::AlwaysNextLine |
+            ElseIfBraceStyle::ClosingNextLine => alt_block_sep,
+            _ => " ",
```
Comment (marcusklaas) [inline src/expr.rs]: Nittiest of nits: there's only one case left here, right? I think it's more idiomatic to provide that instead of `_`.

**Comment by marcusklaas** [comment]
This is shaping up nicely! There's a few minor things left before this is ready to merge. In addition to the nits, could you also implement the new option for `match`? Also, could you squash some of the commits? I think having two commits, one for the if-else logic and one for the rest would be ideal.

Finally -- and totally optionally! --, when the brace is placed on the next line, the budget for the conditional is increased by two characters. Could you implement this? Shouldn't be too hard: we simply make the subtraction here conditional [here](https://github.com/rust-lang-nursery/rustfmt/pull/948/files#diff-180bb3ad456cc10517d878fb7795341cR715) conditional.

Thanks for your efforts so far!

**Comment by flutterhash** [comment]
I would also like some input on how to handle the alt_block_sep allocation: currently, every recursive call to rewrite_if_else, and every call to Loop.rewrite(...) allocate the same string unnecessarily. Ideally, it'd just be allocated once at the start of the Expr rewrite function and then passed down via an input parameter. rewrite_if_else already seems quite busy with input arguments and so I am hesitant.

**Comment by marcusklaas** [comment]
I noticed that too, but I wouldn't worry about it. A super small allocation for every if-else is a small price compared to the added complexity of having another function argument. I'd leave it as is.

**Comment by flutterhash** [comment]
Okay, I'll do the budget adjustment as well. Should the match and loop styles be merged together into one option?

**Comment by marcusklaas** [comment]
Cool! One option would be perfect :+1:

**Comment by flutterhash** [comment]
The changes have been made. Thank you for your patience with me and the embarassingly poor quality of the initial submission. I'm still a bit unhappy with the amount of repetition that there is, but I don't wish to be too invasive with the changes.

**Comment by marcusklaas** [comment]
Don't worry about it, this was an excellent first PR :-) Thanks again!

**Pull request merged**